### PR TITLE
Make site work even without localStorage

### LINKED ISSUE #6758: All pages on MDN Web Docs are blank on seamonkey browser
### Summary

When trying to access any of the pages with javascript enabled I briefly see the page, and then it is replaced with blank page. 

The javascript console has following errors:

Content Security Policy: Couldn’t parse invalid host 'report-sample'
Content Security Policy: Couldn’t process unknown directive ‘script-src-elem’
Content Security Policy: Couldn’t parse invalid host 'report-sample'
SecurityError: The operation is insecure.
Not_a_function:3
sessionStorage.getItem didn't work SecurityError: The operation is insecure.
user-context.tsx:60:12
DOMException
code: 18
columnNumber: 0
data: null
filename: "https://developer.mozilla.org/static/js/main.38618567.js"
lineNumber: 2
message: "The operation is insecure."
name: "SecurityError"
result: 2152923154
stack: "Le/<@https://developer.mozilla.org/static/js/main.38618567.js:2:363279\nel@https://developer.mozilla.org/static/js/main.38618567.js:2:230855\nws@https://developer.mozilla.org/static/js/main.38618567.js:2:251273\nxs/</<@https://developer.mozilla.org/static/js/main.38618567.js:2:247914\nD@https://developer.mozilla.org/static/js/main.38618567.js:2:298349\nN@https://developer.mozilla.org/static/js/main.38618567.js:2:298883\n"
__proto__: DOMExceptionPrototype { name: Getter, message: Getter, INDEX_SIZE_ERR: 1, … }
react-dom.production.min.js:186:401
DOMException
code: 18
columnNumber: 0
data: null
filename: "https://developer.mozilla.org/static/js/main.38618567.js"
lineNumber: 2
message: "The operation is insecure."
name: "SecurityError"
result: 2152923154
stack: "y/<@https://developer.mozilla.org/static/js/main.38618567.js:2:3456\nel@https://developer.mozilla.org/static/js/main.38618567.js:2:230855\nws@https://developer.mozilla.org/static/js/main.38618567.js:2:251273\nxs/</<@https://developer.mozilla.org/static/js/main.38618567.js:2:247914\nD@https://developer.mozilla.org/static/js/main.38618567.js:2:298349\nN@https://developer.mozilla.org/static/js/main.38618567.js:2:298883\n"
__proto__: DOMExceptionPrototype { name: Getter, message: Getter, INDEX_SIZE_ERR: 1, … }
react-dom.production.min.js:186:401
GETXHR
https://developer.mozilla.org/api/v1/whoami
SecurityError: The operation is insecure.
main.38618567.js:2

### URL

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Not_a_function

### Reproduction steps

1. Download seamonkey from https://www.seamonkey-project.org/start/
2. Install it and start it
3. go to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Not_a_function page

### Expected behavior

To see page normally

### Actual behavior

Not seeing anything, and getting security errors on the console.

### Device

Laptop

### Browser

Other (specify below)

### Browser version

Stable

### Operating system

Linux

### Screenshot

image

### Anything else?

My browser is seamonkey version 2.53.13.

### Validations

- [X] I have read the [Community Participation Guidelines](https://www.mozilla.org/en-US/about/governance/policies/participation/).
- [X] I have verified that there isn't [already an issue](https://github.com/mdn/yari/issues) that reports the same bug to avoid creating a duplicate.
- [X] I have checked that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mdn/yari/discussions).

### PR DESCRIPTION
**Pull request by bershanskiy**
## Summary

Fixes #6758.

Make the site work even when the user disables cookies completely (some features are not available, e.g., remembering color theme preference, frequently viewed documents, recently viewed documents.

### Problem
Previously scripts were calling `localStorage.setItem` and `localStorage.getItem` outside of try-catch and were crashing.

### Solution
Wrap each call to `localStorage.setItem` and `localStorage.getItem` into try-catch.

---

## How did you test this change?
1. Opened the site and saw that it loads properly
2. Changed theme and saw that page color changes (but gets reset upon navigation and reload)
3. Opened site root with cookies disabled and saw a message that the recently viewed document list is unavailable without cookies.

### CODE REVIEW

File: client/src/writers-homepage/viewed-documents.tsx [hunk 1]
```diff
@@ -46,8 +57,10 @@ export default function ViewedDocuments() {
       aria-labelledby="recently-viewed-documents"
     >
       <h3>Recently viewed documents</h3>
-      {!entries ? (
+      {entries === undefined ? (
         <Banner>Loading recently viewed documents</Banner>
+      ) : entries === null ? (
+        <Banner>This feature requires cookies</Banner>
```
Comment (bershanskiy) [inline client/src/writers-homepage/viewed-documents.tsx:63]: Here I can add a more specific message like this:
```
: entries === null ? (
  navigator.cookieEnabled ?
    (<Banner>Could not be loaded for unknown reason</Banner>) : 
    (<Banner>Could not be loaded, try enabling first-party cookies for this site</Banner>))
```

File: client/src/document/hooks.ts [hunk 2]
```diff
@@ -154,16 +179,8 @@ export function useFrequentlyViewed(): [
   }, [updated]);
 
   const setStoredEntries = (value: FrequentlyViewedEntry[]) => {
-    try {
-      setEntries(entries);
```
[the remark this answers is not in the document]
Comment (caugner) [reply client/src/document/hooks.ts:165]: Yup, should probably be `value`, like you suggested. 👍

**Comment by bershanskiy** [comment]
I'll sign the commit now.

**Comment by bershanskiy** [comment]
Is this PR relevant? Can I do anything to advance it?

**Comment by bershanskiy** [comment]
Should I do anything to advance this PR?

**Comment by bershanskiy** [comment]
@caugner Is this PR still relevant?

**Review by caugner — CHANGES_REQUESTED** [review]
One consistency nit, otherwise looking good. 👍 

Could you please rebase your branch onto the latest main, and I will test this once more before merging?

**Comment by bershanskiy** [comment]
@caugner Thank for the review! I fixed all the style issues and made sure the commit has a valid signature.

**Comment by bershanskiy** [comment]
@caugner Should I do anything else to advance this PR?

**Comment by caugner** [comment]
I have triggered a dev build, which is now deployed here: https://pr6352.content.dev.mdn.mozit.cloud/

Will hopefully find time to test these days.

**Comment by bershanskiy** [comment]
Thanks for publishing dev build, I tested that and everything worked as expected: the content appears, but I can not log in and most setting changes do not persist after page navigations and reloads. In particular:
 - interactive demos work
 - color theme chooser works, but does not persist
 - language selection works (and persists on page reloads, as expected because it is a part of URL)
 - search works
 - landing page sections "Recent contributions" and "Latest news" work
 - MDN Plus promo banner reappears on every reload
 - MDN Plus button "Already a subscriber?" does nothing meaningful
 - MDN Plus button "Get MDN Plus" opens the correct page, but there is only one plan ("Free"/"Core") and it is impossible to sign up for it

When I filed this PR in May, I wrote that "some features are not available, e.g., remembering color theme preference, frequently viewed documents, recently viewed documents". Of course, this is still the case and I planned to implement some of these features in a follow-up PR. I still think we should merge this as is and if someone asks implement these features.

File: client/src/document/toolbar/index.tsx [hunk 3]
```diff
@@ -17,20 +17,30 @@ export default function Toolbar({
...
+        entry,
+        ...previousVisits.filter((v) => v.url !== entry.url),
+      ];
+      localStorage.setItem(
+        localStorageKey,
+        JSON.stringify(visits.slice(0, 20))
+      );
+    } catch (err) {
+      console.warn(
+        "Unable to update list of viewed documents in localStorage",
```
Comment (caugner) [inline client/src/document/toolbar/index.tsx:40]:
```suggestion
        "Unable to write viewed documents to localStorage",
```

File: client/src/settings/mdn-worker.tsx [hunk 4]
```diff
@@ -124,10 +124,16 @@ export class MDNWorker {
...
-      JSON.parse(window.localStorage.getItem("MDNSettings") || "null") ??
-      new SettingsData()
-    );
+    let settingsData: SettingsData | null = null;
+    try {
+      settingsData = JSON.parse(
+        window.localStorage.getItem("MDNSettings") || "null"
+      );
+    } catch (err) {
+      console.warn("Unable to retreive MDNSettings from localStorage", err);
```
Comment (caugner) [inline client/src/settings/mdn-worker.tsx:133]:
```suggestion
      console.warn("Unable to read settings from localStorage", err);
```

File: client/src/settings/mdn-worker.tsx [hunk 5]
```diff
@@ -165,7 +171,11 @@ export class MDNWorker {
     }
 
     const settings = { ...current, ...settingsData };
-    window.localStorage.setItem("MDNSettings", JSON.stringify(settings));
+    try {
+      window.localStorage.setItem("MDNSettings", JSON.stringify(settings));
+    } catch (err) {
+      console.warn("Unable to save MDNSettings to localStorage", err);
```
Comment (caugner) [inline client/src/settings/mdn-worker.tsx:177]:
```suggestion
      console.warn("Unable to write settings to localStorage", err);
```

File: libs/constants/index.js [hunk 6]
```diff
@@ -76,9 +76,9 @@ const scriptSrcValues = [
    * we must always update the CSP hash (see instructions there).
    */
   // - Previous hash (to avoid cache invalidation issues):
-  "'sha256-GA8+DpFnqAM/vwERTpb5zyLUaN5KnOhctfTsqWfhaUA='",
-  // - Current hash:
   "'sha256-R1d+R+j7z1re3EpSIuqRpflohOvTm3fAxKvfEiXgC1o='",
```
Comment (caugner) [inline libs/constants/index.js:79]: Keep previous hash, as it already is the previous hash (from last night's build):

```suggestion
  "'sha256-GA8+DpFnqAM/vwERTpb5zyLUaN5KnOhctfTsqWfhaUA='",
```

**Review by caugner — CHANGES_REQUESTED** [review]

**Review by caugner — APPROVED** [review]

**Pull request merged**